### PR TITLE
fix: and refactor geometric digitization

### PR DESCRIPTION
**Pull request by benjaminhuth**
* fix: activation was only 2D path, but must be 3D path. Otherwise, an upright track would get 0 activation...
* refactor: rename `Channelizer` to `Segmentizer`, add `Channelizer` that ties together the things
* added unit test

### CODE REVIEW

**Comment by asalzburger** [comment]
Unit test is failing - is that expected ? Do we need to update the test?

**Comment by benjaminhuth** [comment]
Hmm no is not expected. They *should* run through. I will check after the workshop

**Comment by AJPfleger** [comment]
Could you also adapt the tests to what I did in #2632 ? :)

I didn't touch `Tests/UnitTests/Fatras/Digitization/ChannelizerTests.cpp` in my PR to avoid conflicts on your side.
Also the file `Tests/UnitTests/Fatras/Digitization/SegmentizerTests.cpp` will be new for me.

File: Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationAlgorithm.hpp [hunk 1]
```diff
@@ -93,7 +78,7 @@ class DigitizationAlgorithm final : public IAlgorithm {
   /// @return the list of digitized parameters
   DigitizedParameters localParameters(
       const GeometricConfig& geoCfg,
-      const std::vector<ActsFatras::Channelizer::ChannelSegment>& channels,
+      const std::vector<ActsFatras::Segmentizer::ChannelSegment>& channels,
```
Comment (asalzburger) [inline Examples/Algorithms/Digitization/include/ActsExamples/Digitization/DigitizationAlgorithm.hpp:81]: Makes sense! The naming is clearer.

File: Examples/Algorithms/Digitization/src/DigitizationAlgorithm.cpp [hunk 2]
```diff
@@ -205,16 +205,20 @@ ActsExamples::ProcessCode ActsExamples::DigitizationAlgorithm::execute(
...
                            << " parameters.");
-              auto channels = channelizing(digitizer.geometric, simHit,
-                                           *surfacePtr, ctx.geoContext, rng);
-              if (channels.empty()) {
+              const auto& cfg = digitizer.geometric;
+              Acts::Vector3 driftDir = cfg.drift(simHit.position(), rng);
+              auto channelsRes = m_channelizer.channelize(
+                  simHit, *surfacePtr, ctx.geoContext, driftDir,
+                  cfg.segmentation, cfg.thickness);
+              if (!channelsRes.ok() || channelsRes->empty()) {
```
Comment (asalzburger) [inline Examples/Algorithms/Digitization/src/DigitizationAlgorithm.cpp:213]: I like the change to the Result object as well.

File: Fatras/include/ActsFatras/Digitization/Channelizer.hpp [hunk 3]
```diff
@@ -8,97 +8,66 @@
...
+  ///
+  /// @param hit The hit we want to channelize
+  /// @param surface the surface on which the hit is
+  /// @param gctx the Geometry context
+  /// @param driftDir the drift direction
+  /// @param segmentation the segmentation of the surface
+  /// @param thickness the thickness of the surface
+  ///
+  /// TODO Why having the thickness as a config parameter,
+  /// and not pulling it from the surface->detectorElement directly?
```
Comment (asalzburger) [inline Fatras/include/ActsFatras/Digitization/Channelizer.hpp:37]: Actually, we might want to remove the "thickness" from the detector element as it is a very planar detector oriented concept.

File: Fatras/include/ActsFatras/Digitization/Channelizer.hpp [hunk 4]
```diff
@@ -8,97 +8,66 @@
...
 
-    /// Constructor with arguments
-    ///
-    /// @param bin_ The bin corresponding to this step
-    /// @param path2D_ The start/end 2D position of the segement
-    /// @param activation_ The segment activation (clean: length) for this bin
-    ChannelSegment(Bin2D bin_, Segment2D path2D_, double activation_)
-        : bin(bin_), path2D(std::move(path2D_)), activation(activation_) {}
-  };
+    // Go from 2D-path to 3D-path by applying thickness
```
Comment (asalzburger) [inline Fatras/include/ActsFatras/Digitization/Channelizer.hpp:57]: I don't see where the thickness is applied here.

File: Fatras/include/ActsFatras/Digitization/Channelizer.hpp [hunk 5]
```diff
@@ -8,97 +8,66 @@
...
-  /// @param surface The surface for the channelizing
-  /// @param segmentation The segmentation for the channelizing
-  /// @param segment The surface segment (cartesian coordinates)
-  ///
-  /// @return a vector of ChannelSegment objects
-  std::vector<ChannelSegment> segments(const Acts::GeometryContext& geoCtx,
-                                       const Acts::Surface& surface,
-                                       const Acts::BinUtility& segmentation,
-                                       const Segment2D& segment) const;
+    for (auto& seg : segments) {
```
Comment (asalzburger) [inline Fatras/include/ActsFatras/Digitization/Channelizer.hpp:62]: Forget the comment above, it's clear now.

File: Fatras/include/ActsFatras/Digitization/Channelizer.hpp [hunk 6]
```diff
@@ -8,97 +8,66 @@
...
+      auto r = path2D != 0.0 ? (seg.activation / path2D) : 1.0;
+      auto segThickness = r * thickness;
+
+      seg.activation = std::hypot(segThickness, seg.activation);
```
Comment (asalzburger) [inline Fatras/include/ActsFatras/Digitization/Channelizer.hpp:66]: Is there a unit test for this?
I.e. that the 3D path length from the segments add up to the full 3D path after segmentation?

File: Tests/UnitTests/Fatras/Digitization/ChannelizerTests.cpp [hunk 7]
```diff
@@ -1,218 +1,143 @@
...
+BOOST_AUTO_TEST_CASE(test_upright_particle) {
+  Helper helper;
+
+  Acts::Vector3 pos3 = Acts::Vector3{10_um, 10_um, 0.0};
+  Acts::Vector3 dir3 = Acts::Vector3{0.0, 0.0, helper.thickness}.normalized();
+
+  auto segments = helper.channelize(pos3, dir3);
+
+  BOOST_CHECK(segments.size() == 1);
+  BOOST_CHECK_CLOSE(segments[0].activation, helper.thickness, 1.e-8);
```
Comment (asalzburger) [inline Tests/UnitTests/Fatras/Digitization/ChannelizerTests.cpp:68]: This is right through, I see.

File: Tests/UnitTests/Fatras/Digitization/ChannelizerTests.cpp [hunk 8]
```diff
@@ -1,218 +1,143 @@
...
+  const double disp = 10_um;
+
+  Acts::Vector3 hitPosition = Acts::Vector3{10_um, 10_um, 0.0};
+  Acts::Vector3 hitDirection =
+      Acts::Vector3({disp, 0.0, helper.thickness}).normalized();
+
+  auto segments = helper.channelize(hitPosition, hitDirection);
+
+  BOOST_CHECK(segments.size() == 1);
+  BOOST_CHECK_CLOSE(segments[0].activation, std::hypot(disp, helper.thickness),
```
Comment (asalzburger) [inline Tests/UnitTests/Fatras/Digitization/ChannelizerTests.cpp:83]: Next check: done. Tiltet through in one step.

**Review by asalzburger — DISMISSED** [review]
Makes sense, let's fix the UnitTests and get that in.

**Comment by benjaminhuth** [comment]
@asalzburger The PR should now be ready, can you approve?

File: Tests/UnitTests/Fatras/Digitization/SegmentizerTests.cpp [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+  Acts::Vector2 sPositionE(7.03, -0.3);
+  auto sSegment =
+      cl.segments(geoCtx, *radialDisc, strips, {sPositionS, sPositionE});
+  BOOST_CHECK(sSegment.size() == 59);
+
+  // Test: jump over R boundary, but stay in phi bin
+  sPositionS = Acts::Vector2(6.76, 0.);
+  sPositionE = Acts::Vector2(7.83, 0.);
+  sSegment = cl.segments(geoCtx, *radialDisc, strips, {sPositionS, sPositionE});
+  BOOST_CHECK(sSegment.size() == 2);
```
Comment (AJPfleger) [inline Tests/UnitTests/Fatras/Digitization/SegmentizerTests.cpp:118]:
```suggestion
  BOOST_CHECK_EQUAL(nSegments.size(), 1);
  BOOST_CHECK_EQUAL(nSegments[0].bin[0], 0);
  BOOST_CHECK_EQUAL(nSegments[0].bin[1], 161);

  // Test: now opver more phi strips
  Acts::Vector2 sPositionS(6.76, 0.5);
  Acts::Vector2 sPositionE(7.03, -0.3);
  auto sSegment =
      cl.segments(geoCtx, *radialDisc, strips, {sPositionS, sPositionE});
  BOOST_CHECK_EQUAL(sSegment.size(), 59);

  // Test: jump over R boundary, but stay in phi bin
  sPositionS = Acts::Vector2(6.76, 0.);
  sPositionE = Acts::Vector2(7.83, 0.);
  sSegment = cl.segments(geoCtx, *radialDisc, strips, {sPositionS, sPositionE});
  BOOST_CHECK_EQUAL(sSegment.size(), 2);
```

**Review by AJPfleger — CHANGES_REQUESTED** [review]

**Review by asalzburger — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [a01f82a854d0ce09ed29d739185d3044788dc50a]
### Build job with this PR failed!
Please investigate the [build job](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/33897580) for the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/6472121)!

**Pull request merged**